apps/dcos_net/src/dcos_net_node.erl

```diff
@@ -0,0 +1,258 @@
+-module(dcos_net_node).
+-behaviour(gen_server).
+
+%% API
+-export([
+    start_link/0,
+    dump/0,
+    force_dump/0
+]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2,
+    handle_info/2, terminate/2, code_change/3]).
+
+-export_type([metadata/0]).
+
+-type metadata() :: #{
+    public_ips => [inet:ip_address()],
+    hostname => binary(),
+    updated => non_neg_integer()
+}.
+
+-record(state, {
+    timer_ref :: reference(),
+    lashup_ref :: reference(),
+    refresh = #{} :: map()
+}).
+-type state() :: #state{}.
+
+-include_lib("stdlib/include/ms_transform.hrl").
+-define(LASHUP_KEY, [dcos_net, nodes]).
+
+-spec(start_link() ->
+    {ok, Pid :: pid()} | ignore | {error, Reason :: term()}).
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+-spec(dump() -> #{inet:ip4_address() => metadata()}).
+dump() ->
+    List = lashup_kv:value(?LASHUP_KEY),
+    dump(List).
+
+-spec(force_dump() -> #{inet:ip4_address() => metadata()}).
+force_dump() ->
+    % It's not possible to unsubscribe from lashup_kv_events_helper's updates.
+    % force_dump/0 function starts a new process that does all the work, sends
+    % the result back and terminates itself.
+    % TODO: Add ubsibscribe to the events helper and get rid of spawn_link.
+    Ref = make_ref(),
+    PPid = self(),
+    proc_lib:spawn_link(fun() -> force_refresh(PPid, Ref) end),
+    Timeout = force_refresh_timeout(),
+    receive
+        {Ref, Result} ->
+            Result
+    after Timeout ->
+        exit(timeout)
+    end.
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+init([]) ->
+    self() ! init,
+    {ok, []}.
+
+handle_call(_Request, _From, State) ->
+    {noreply, State}.
+
+handle_cast(_Request, State) ->
+    {noreply, State}.
+
+handle_info(init, []) ->
+    LRef = subscribe(),
+    RTerm = get_refresh_value(),
+    TRef = start_update_timer(0),
+    {noreply, #state{timer_ref=TRef, lashup_ref = LRef, refresh = RTerm}};
+handle_info({timeout, TRef, update}, State = #state{timer_ref = TRef}) ->
+    ok = update_node_info(),
+    TRef0 = start_update_timer(),
+    {noreply, State#state{timer_ref=TRef0}, hibernate};
+handle_info({lashup_kv_events, #{ref := Ref, value := Value}},
+            State = #state{lashup_ref = Ref}) ->
+    {noreply, handle_kv_event(Value, State)};
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+-spec(start_update_timer() -> reference()).
+start_update_timer() ->
+    Timeout = application:get_env(
+        dcos_net, update_node_info, timer:minutes(30)),
```

Comment:
have you considered the tradeofs of making this configurable form day 0? I wonder if we would want to be able to tweak and test this time without touching code.

Reply:
Maybe same way we did with `detect_ip_public`

Reply:
`detect_ip_public` runs locally and doesn't depend on the size of a cluster, so we can use a reasonable constant as its timeout. On big clusters force refresh can take more time than on small ones. We don't have the goal to have 100% code coverage, force refresh and this timer run exactly the same code.